File: apple_watch_app/PathOfNurWatch/Notifications/PrayerReminderPolicy.swift
import Foundation

struct PrayerReminderPolicy {
    func shouldScheduleInitial(
        prayer: WatchPrayerStatus,
        now: Date,
        settings: WatchNotificationSettings,
        dedupStates: [NotificationDedupState]
    ) -> Bool {
        guard settings.prayerNotificationsEnabled,
              settings.enabledPrayerIds.contains(prayer.prayerId),
              prayer.status == .pending,
              prayer.scheduledTime > now,
              !settings.quietModeEnabled else {
            return false
        }
        let key = dedupKey(prayerId: prayer.prayerId, kind: .initial, scheduledAt: prayer.scheduledTime)
        return !hasLiveDedup(key: key, in: dedupStates)
    }

    func shouldScheduleFollowUp(
        prayer: WatchPrayerStatus,
        baseTime: Date,
        now: Date,
        settings: WatchNotificationSettings,
        dedupStates: [NotificationDedupState]
    ) -> Bool {
        guard settings.prayerNotificationsEnabled,
              settings.followUpReminderEnabled,
              settings.enabledPrayerIds.contains(prayer.prayerId),
              prayer.status == .pending else {
            return false
        }
        let followUpTime = Calendar.current.date(byAdding: .minute, value: settings.followUpDelayMinutes, to: baseTime) ?? baseTime
        guard followUpTime > now else { return false }
        let key = dedupKey(prayerId: prayer.prayerId, kind: .followUp, scheduledAt: followUpTime)
        return !hasLiveDedup(key: key, in: dedupStates)
    }

    func shouldAllowSnooze(
        prayerId: String,
        snoozeAt: Date,
        dedupStates: [NotificationDedupState]
    ) -> Bool {
        let key = dedupKey(prayerId: prayerId, kind: .snooze, scheduledAt: snoozeAt)
        return !hasLiveDedup(key: key, in: dedupStates)
    }

    func shouldCancel(prayer: WatchPrayerStatus) -> Bool {
        prayer.status == .completed
    }

    func dedupKey(prayerId: String, kind: WatchNotificationKind, scheduledAt: Date) -> String {
        let stamp = ISO8601DateFormatter().string(from: scheduledAt)
        return "\(prayerId)|\(kind.rawValue)|\(stamp)"
    }

    private func hasLiveDedup(key: String, in states: [NotificationDedupState]) -> Bool {
        states.contains(where: { $0.dedupKey == key && $0.lastCancelledAt == nil })
    }
}

File: wear_os_app/app/src/main/java/com/pathofnur/watch/notifications/PrayerReminderPolicy.kt
package com.pathofnur.watch.notifications

import com.pathofnur.watch.domain.NotificationDedupState
import com.pathofnur.watch.domain.WatchNotificationKind
import com.pathofnur.watch.domain.WatchNotificationSettings
import com.pathofnur.watch.domain.WatchPrayerStatus
import com.pathofnur.watch.domain.PrayerState
import java.time.Instant

class PrayerReminderPolicy {
    fun shouldScheduleInitial(
        prayer: WatchPrayerStatus,
        scheduledAt: Instant,
        now: Instant,
        settings: WatchNotificationSettings,
        dedupStates: List<NotificationDedupState>
    ): Boolean {
        if (!settings.prayerNotificationsEnabled || settings.quietModeEnabled) return false
        if (prayer.status == PrayerState.Completed) return false
        if (!settings.enabledPrayerIds.contains(prayer.prayerId)) return false
        if (!scheduledAt.isAfter(now)) return false
        return !hasLiveDedup(dedupKey(prayer.prayerId, WatchNotificationKind.Initial, scheduledAt), dedupStates)
    }

    fun shouldScheduleFollowUp(
        prayer: WatchPrayerStatus,
        followUpAt: Instant,
        settings: WatchNotificationSettings,
        now: Instant,
        dedupStates: List<NotificationDedupState>
    ): Boolean {
        if (!settings.prayerNotificationsEnabled || !settings.followUpReminderEnabled) return false
        if (prayer.status == PrayerState.Completed) return false
        if (!settings.enabledPrayerIds.contains(prayer.prayerId)) return false
        if (!followUpAt.isAfter(now)) return false
        return !hasLiveDedup(dedupKey(prayer.prayerId, WatchNotificationKind.FollowUp, followUpAt), dedupStates)
    }

    fun shouldAllowSnooze(prayerId: String, snoozeAt: Instant, dedupStates: List<NotificationDedupState>): Boolean {
        return !hasLiveDedup(dedupKey(prayerId, WatchNotificationKind.Snooze, snoozeAt), dedupStates)
    }

    fun dedupKey(prayerId: String, kind: WatchNotificationKind, scheduledAt: Instant): String {
        return "$prayerId|${kind.name}|${scheduledAt.toEpochMilli()}"
    }

    private fun hasLiveDedup(key: String, states: List<NotificationDedupState>): Boolean {
        return states.any { it.dedupKey == key && it.lastCancelledAt == null }
    }
}

